### PR TITLE
fix(acp): reliable structured-view stop, turn-end, and session recovery

### LINKED ISSUE #2237: Structured view: Stop button is a no-op when the daemon has no prompt in flight, stuck spinner only escapes via aoe acp restart
**Note:** Claude handled the investigation and prose via back-and-forth. Idea, decisions, and everything else are mine. — @Seluj78

## Problem

Structured-view sessions get stuck "thinking" / mid-tool with a spinner that never clears, and the composer Stop button does nothing. The only escape is `aoe acp restart <id>`. This is happening more and more and makes long-lived sessions unusable.

The Stop button is not just slow here, it is a hard no-op: clicking it once, twice, or ten times changes nothing. The agent appears frozen in a random state (thinking, or partway through writing a file), and the turn never ends.

## Root cause

This is the acknowledged residual of #1216, now reachable through the everyday Stop button.

The daemon's per-turn prompt loop owns the terminal `Stopped` and the cancel-escalation path. When a worker reattaches or respawns and adopts an already in-flight turn, that turn has no owning `prompt_fut` on the new client. The resume-idle watchdog that would synthesize a `Stopped` for an orphaned turn disarms permanently the moment the adopted turn emits its first event:

`src/acp/acp_client.rs:4634-4650`
```rust
if first_event_after_attach.load(Ordering::Relaxed) {
    // ... any further silence is normal mid-turn reasoning ... Disarm
    // permanently. The narrow residual (the turn completes after attach
    // and its PromptResponse is lost, leaving a stale spinner) is rare
    // and recoverable via force-end-turn / a new prompt. See #1216.
    return;
}
```

So when the adopted turn finishes (or wedges) and its `PromptResponse` is lost, no terminal `Stopped` is ever emitted. The UI keeps `turnActive = true` while the daemon has no prompt in flight. This is the "rare" residual, except it is no longer rare: every build-stale respawn adopts the in-flight turn, and someone developing `aoe` rebuilds constantly. Session `768e497afa9e4415`'s worker log shows the exact sequence (build-stale respawns adopting an in-flight turn, then a Stop that finds no prompt, then the manual restart):

```
05:43:57  adopting build-stale structured view worker to drain in-flight turn before respawn
06:01:59  sending session/cancel (no prompt in flight)
06:02:09  registry entry gone while worker handle live; tearing down reason=restart_pending
```

Now the second failure, which is what makes it unrecoverable from the UI: with no prompt in flight, the Stop button is structurally a no-op.

Layer 1, backend. With no in-flight prompt, `ClientCmd::Cancel` sends a bare `session/cancel` notification and nothing else, no synthetic `Stopped`, no respawn:

`src/acp/acp_client.rs:5237-5240`
```rust
Some(ClientCmd::Cancel) => {
    info!(target: "acp.protocol", "sending session/cancel (no prompt in flight)");
    connection.send_notification(CancelNotification::new(acp_session_id.clone()))?;
}
```

Layer 2, web. The Stop button only escalates to force-stop when `acp.state.cancelling` is true:

`web/src/components/acp/AcpRuntime.tsx:179-190`
```tsx
onCancel: async () => {
  if (acp.state.cancelling) {
    await acp.forceEndTurn();
  } else {
    await acp.cancelPrompt();
  }
}
```

But `cancelling` flips true only on a `CancelRequested` event (`web/src/lib/acpTypes.ts:642-647`), and that event is emitted only from inside the in-flight prompt loop (`src/acp/acp_client.rs:5001-5003`). With no prompt in flight there is no `CancelRequested`, so `cancelling` stays false forever, so every Stop click takes the `cancelPrompt()` branch. The user can never reach `forceEndTurn()` through the button. `force_end_turn` is the one path that publishes a synthetic `Stopped` and respawns the worker (`src/acp/supervisor.rs:1987-1999`), and it is unreachable.

Net: UI says active, daemon says idle, the escape hatch the #1216 comment points at ("recoverable via force-end-turn") cannot be triggered from the UI, and `aoe acp restart` is the only way out.

## Evidence from the affected sessions

`577a11d9e8374da4` (the chronic one, 22813 events, weeks old). Stop pressed during in-flight prompts 16 times, silent-orphan watchdog fired 4 times (all with `in_flight_tools=0`), one force-stop, many `restart_pending`:

```
prompt_orphaned (x4), restart_pending (x6), user_stopped (x3), user_forced (x1), cancelled (many)
```

Every silent-orphan fire logged `in_flight_tools=0`. The "stuck writing a file" variant (a tool call started and never completed) is never caught, because the silent-orphan watchdog is suppressed whenever a tool is in flight by design:

`src/acp/acp_client.rs` `SilentOrphanWatchdog::should_fire` requires `self.tool_calls_in_flight.is_empty()`. A `ToolCallStarted` with no matching `ToolCallCompleted` keeps the set non-empty forever, so the watchdog stays parked and the turn never gets a synthesized `Stopped`.

`768e497afa9e4415`: build-stale respawns adopting the in-flight turn, then `session/cancel (no prompt in flight)`, then manual restart (quoted above).

(Session `0868119de7d24a4f` turned out to be this very investigation session, healthy, not a hang. Disregard it.)

### A. Make the Stop button always escalate, do not depend on the server `cancelling` flag (web, minimum fix)

Track the cancel attempt locally so a second click always force-ends, even when no `CancelRequested` ever arrives:

```tsx
const cancelRequestedRef = useRef(false);
onCancel: async () => {
  if (acp.state.cancelling || cancelRequestedRef.current) {
    cancelRequestedRef.current = false;
    await acp.forceEndTurn();
  } else {
    cancelRequestedRef.current = true;
    await acp.cancelPrompt();
  }
}
```
Reset `cancelRequestedRef` on any `Stopped` / new prompt. This gives the user a working escape (second click respawns the worker and resyncs) without `aoe acp restart`, regardless of whether the daemon has a prompt in flight.

### B. Emit a terminal `Stopped` when Cancel finds no prompt in flight (backend, root resync)

The no-prompt-in-flight `ClientCmd::Cancel` arm (`acp_client.rs:5237`) should publish a synthetic `Stopped { reason: "cancelled" }` on the event channel after the bare notification. By definition no turn is in flight, so emitting a terminal event is correct and resyncs the UI to the daemon's truth on the first click. This makes the desync self-healing instead of permanent.

### C. Keep a lightweight completion watchdog for adopted orphan turns (backend, prevents the desync)

The #1216 disarm-permanently decision is the upstream cause. Instead of disarming the resume-idle watchdog forever on first post-attach event, keep a reduced-grace orphan-completion watchdog armed for adopted turns so a turn that completes (or wedges) with a lost `PromptResponse` still gets a synthesized terminal `Stopped`. Also consider treating a long-lived `ToolCallStarted` with no `ToolCallCompleted` as a wedge candidate (the current `tool_calls_in_flight.is_empty()` gate makes a stuck tool call invisible forever).

A alone restores usability. B + C remove the desync so the spinner clears without any click.

## Why now

- Core flow, reported as "unusable". A stuck spinner with a dead Stop button is the worst failure shape for the structured view.
- Hits developers of `aoe` hardest (frequent rebuilds trigger build-stale respawns that adopt in-flight turns), which is exactly the maintainer's daily loop.
- The escape the code comment promises ("recoverable via force-end-turn") is unreachable from the UI; the fix is small and high-leverage.

## User stories / regression tests

Implementation MUST add tests covering each story below; PR is not done without them.

- As a user whose turn is stuck with no prompt in flight on the daemon, when I click Stop a second time, then the turn force-ends (worker respawns, spinner clears) without my running `aoe acp restart`.
  - Suggested test layer: Vitest + RTL on the `AcpRuntime` `onCancel` escalation (assert second click calls `forceEndTurn` when `cancelling` is false), plus `web/tests/live/` for the end-to-end unstick.
- As a user, when the daemon receives a cancel with no prompt in flight, then a terminal `Stopped` is published and `turnActive` returns to false.
  - Suggested test layer: Rust unit/integration on the `ClientCmd::Cancel` no-prompt-in-flight arm in `src/acp/acp_client.rs`.
- As a user whose adopted turn completes after a build-stale respawn, when its `PromptResponse` is lost, then a synthesized terminal `Stopped` still clears the spinner within the watchdog grace.
  - Suggested test layer: Rust unit on the orphan-completion watchdog (drive an adopted turn that emits events then goes silent without a PromptResponse).

## Open questions

- Proposal B emits `Stopped { reason: "cancelled" }` unconditionally in the outer cmd loop. Between turns this is harmless (UI already idle), but confirm no consumer treats an extra `cancelled` while idle as meaningful.
- Should C reuse `SilentOrphanWatchdog` with a separate "adopted turn" config, or a distinct lighter watchdog, to avoid regressing the #1176 / #1858 grace tuning?
- Is the stuck-tool-in-flight variant (no `ToolCallCompleted`) common enough to warrant its own grace, or does B's manual escape cover it adequately?

## Out of scope

- Fixing claude-agent-acp emitting session updates after resolving `session/prompt` (upstream adapter behavior).
- The opencode-specific cancel-ignored case already tracked in #1908.
- Client-side composer queueing of prompts during a stuck turn (#1031).

## Cross-references

- #1216 — the resume-idle watchdog disarm-permanently decision whose "rare residual" this issue makes common.
- #1908 (open) — Stop button ineffective with opencode; adjacent but distinct (prompt IS in flight there, adapter ignores cancel; here there is no in-flight prompt and escalation is unreachable).
- #1727 — added the Stop -> Force-stop escalation that this issue shows is unreachable in the no-prompt-in-flight state.
- #1240, #1858, #1176 — silent-orphan watchdog and its grace tuning; Proposal C must not regress these.
- #1196 — force_end_turn unsticking and dropped follow-ups; same family.
- #1100 — spinner stuck after finish via WS/replay race; related desync surface.

### PR DESCRIPTION
**Pull request by Seluj78**
**Note:** Claude handled the implementation and prose via back-and-forth. Idea, decisions, and everything else are mine. — @Seluj78

## Description

Structured-view sessions could get stuck in several ways that all forced the user to fall back to `aoe acp restart`: a spinner that never clears with a Stop button that did nothing, a turn that actually finished getting force-restarted, and a queued follow-up that never sent. This PR fixes four distinct points in the ACP session lifecycle, all surfaced from real stuck sessions.

**1. Stop button always escalates (web).** The composer Stop button only escalated to force-end when the server-confirmed `cancelling` flag was set, but that flag flips only on a `CancelRequested` event the daemon emits while a prompt is in flight. With no in-flight prompt (an adopted/orphaned turn whose terminal `Stopped` was lost, the residual of #1216), `cancelling` never flips, so every click stayed on the graceful-cancel branch and force-end was unreachable. The button now tracks a client-owned "cancel already requested" ref, so a second press always force-ends regardless of the server flag.

**2. Cancel with no prompt in flight self-heals (backend).** That same desync left the daemon's no-prompt-in-flight `ClientCmd::Cancel` arm sending a bare `session/cancel` and nothing else. It now publishes a terminal `Stopped { reason: "cancelled" }` so the first Stop press resyncs the UI. The reducer caps `lastStoppedSeq` at `pendingUserPromptSeq`, so a spurious `Stopped` while already idle is a no-op.

**3. Finished-but-unacked turns end cleanly instead of restarting (backend).** `claude-agent-acp` sometimes finishes a turn (streams the final message and the cost-populated end-of-turn `UsageUpdate`) but never sends the JSON-RPC `PromptResponse`, leaving `prompt_fut` pending. The silent-orphan watchdog then cancelled and force-restarted the worker, showing "Agent finished but didn't notify the daemon" on a turn that actually succeeded. When the watchdog fires on a turn that already emitted its cost-populated usage with no off-protocol work or in-flight tools, the daemon now treats the cost marker as authoritative and ends the turn cleanly as `prompt_complete` (no restart). The genuinely-wedged case (no cost marker) keeps the orphan-restart path, and the off-protocol guard preserves the monitor / async-agent grace behavior of #1360 / #1401 / #1858.

**4. Stale idle-dormant marker no longer strands a queued prompt (backend).** An idle-auto-stopped session gets an `idle_dormant_since` marker (#1689) so the reconciler stops respawning it. Non-user respawn paths (notably the build-stale respawn #1754) could bring a worker back without clearing it; when that worker later exited, the reconciler's `!is_idle_dormant()` resume filter refused to bring the session back, and the web client parked the user's follow-up prompt waiting for a worker that never returned: a deadlock. The daemon now clears `idle_dormant_since` whenever `AcpSessionAssigned` lands (a live worker is by definition not dormant), persisting the clear even on an unchanged-id reattach. `AcpSessionAssigned` fires only on successful worker init, never during the idle reaper's mark-then-shutdown, so there is no race with #1689.

Closes #2237

## PR Type

- [ ] New Feature
- [x] Bug Fix
- [ ] Refactor
- [ ] Documentation
- [ ] Infrastructure / CI

## How to test

- [ ] Open a structured-view session, send a prompt, and while the agent is working click Stop twice; confirm the second click ends the turn (spinner clears) without waiting out a countdown.
- [ ] Get the UI into a stuck spinner with no prompt in flight on the daemon (restart the worker out of band mid-turn). Click Stop once; confirm the spinner clears (single-click self-heal) instead of needing `aoe acp restart`.
- [ ] Let a `claude-agent-acp` turn finish; confirm it does not produce an "Agent finished but didn't notify the daemon. Restarting worker" banner on a completed turn (it ends cleanly).
- [ ] Let a session auto-stop for idle, send a follow-up, and confirm it wakes and runs. Then force a respawn (e.g. version upgrade) and, after the worker later exits, confirm a queued follow-up still brings the session back without `aoe acp restart`.
- [ ] On a normal completed turn, confirm a single Stop during work still does a graceful cancel first (no regression for the common case).

## Checklist

- [x] New and existing tests pass
- [x] Documentation was updated where necessary
- [ ] For UI changes: included screenshot or recording

## Test Coverage Analysis

**Web dashboard / structured view (Playwright user story)**
- [ ] N/A: this PR doesn't change a user-facing dashboard flow
- [x] Added or updated a Playwright spec under `web/tests/` and updated `web/tests/coverage-matrix.json`
- [ ] Skipped a test, because:

**TUI / CLI (e2e test)**
- [x] N/A: this PR doesn't change TUI rendering, a CLI subcommand, or session lifecycle in a way the e2e harness covers
- [ ] Added or updated an e2e test under `tests/e2e/`
- [ ] Skipped a test, because:

## AI Usage

- [ ] No AI was used
- [x] AI was used

**AI Model/Tool used:** Claude (via Claude Code, agent-of-empires `/aoe-implement` skill)

**Any Additional AI Details you'd like to share:**
Investigation (grounded in the live event store + worker logs of several stuck sessions), plan, `/debate` design consultation, and implementation drafted by Claude under my direction. I made the design calls, reviewed each commit, and ran the manual test steps.

- [x] I am an AI Agent filling out this form (check box if true)

---
<a href="https://app.blacksmith.sh/agent-of-empires/codesmith/agent-of-empires/pr/2239"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1784274309&installation_id=139138837&pr_number=2239&repository=agent-of-empires%2Fagent-of-empires&return_to=https%3A%2F%2Fgithub.com%2Fagent-of-empires%2Fagent-of-empires%2Fpull%2F2239&signature=bc40f2bb66819fd99eee1ee082a09a60294389be806e3f8517de13b11c2d5cc4"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

## Summary

This PR fixes four interconnected reliability issues that leave Agent of Empires sessions stuck and requiring manual restart via CLI.

### What Changed

**Documentation Updates:**
- Updated "Stopping a turn" guide to clarify that the Stop button now always escalates to force-end on a second click, even when the daemon has no prompt in flight
- Added troubleshooting entries explaining recovery for sessions stuck with stale "dormant" markers and for agents that finish output but don't send completion acknowledgement

**Backend Session Management:**
- Modified turn-completion handling to distinguish between finished-but-unacknowledged turns (which now end cleanly) vs. genuinely-wedged turns (which follow the orphan-restart path)
- Added logic to emit a synthetic `Stopped` event when cancel is requested with no prompt in flight, resyncing UI and daemon state
- Added clearing of stale `idle_dormant_since` markers whenever a worker comes online, preventing deadlocks in session recovery

**Web-Side Cancel Escalation:**
- Introduced a client-local "cancel requested" tracker so the Stop button can escalate to force-end on a second click without waiting for server confirmation
- Added helper function `nextCancelAction()` to encapsulate escalation logic

**Test Coverage:**
- Added new regression test file validating Stop button escalation behavior across multiple scenarios
- Extended backend unit tests for cost-marker-based completion and terminal stop-reason precedence
- Updated coverage matrix with new high-risk test entry

### What Was Added

- New helper function `nextCancelAction(cancelling, alreadyRequested)` in web component
- New `terminal_stop_reason()` helper in backend for stop-reason precedence logic
- `cost_seen()` accessor on watchdog for detecting cost-populated completion
- Client-local `cancelRequestedRef` state tracking in AcpRuntime
- New regression test file: `AcpRuntime.cancelEscalation.test.ts`
- New test coverage entries in coverage-matrix.json

### What Was Removed

- Inline stop-reason conditionals (centralized into `terminal_stop_reason()`)

### Benefits

**User Experience:**
- Stop button becomes a functional escape route—clicking it twice always restarts the worker without requiring CLI restart
- Spinner clears automatically when daemon receives cancel with no in-flight prompt, reducing user confusion
- Sessions no longer deadlock when idle-auto-stopped and later respawned

**System Reliability:**
- Eliminates desynchronization between UI and daemon state during session recovery
- Prevents unnecessary worker restarts when agents finish normally but don't send completion acknowledgement
- Stops treating stale dormant markers as permanent blockers to session recovery

**Code Quality:**
- Centralizes stop-reason precedence logic for maintainability
- Adds clear separation between finished-but-unacked vs. genuinely-wedged turn scenarios
- Comprehensive test coverage for previously-unhandled edge cases

### CODE REVIEW

File: src/acp/acp_client.rs [hunk 1]
```diff
@@ -6105,6 +6206,98 @@ mod tests {
...
+            wall,
+            cfg,
+        );
+        // WakeupPending clears cost_seen and sets off-protocol work, so
+        // the clean-completion guard (cost_seen && off_protocol none) is
+        // false: a scheduled-wake turn never takes the prompt_complete
+        // shortcut.
+        assert!(!w.cost_seen());
+        assert!(w.off_protocol_work_seen().is_some());
+    }
```
[the remark this answers is not in the document]
Comment (Seluj78) [reply src/acp/acp_client.rs:6272]: Good catch, applied. Added `LifecycleSignal::TerminalUsage` after the wakeup so the test actually exercises the with-cost case, and assert `cost_seen()` is true while `off_protocol_work_seen()` stays set. Fixed in https://github.com/agent-of-empires/agent-of-empires/pull/2239/commits/e9208646

File: web/src/components/acp/AcpRuntime.tsx [hunk 2]
```diff
@@ -130,6 +142,24 @@ export function AcpRuntime({
...
+  // an acknowledgement from the system it is escaping. See #2237.
+  const cancelRequestedRef = useRef(false);
+  // Reset the local cancel intent the moment the turn is no longer
+  // active (a terminal Stopped landed, or a fresh prompt started), so the
+  // next turn starts from a graceful-cancel-first state.
+  useEffect(() => {
+    if (!acp.state.turnActive) {
+      cancelRequestedRef.current = false;
+    }
+  }, [acp.state.turnActive]);
```
[the remark this answers is not in the document]
Comment (Seluj78) [reply web/src/components/acp/AcpRuntime.tsx]: Fixed in https://github.com/agent-of-empires/agent-of-empires/pull/2239/commits/c42c5814. Extracted the logic into a `useCancelEscalation` hook and replaced the turnActive-keyed reset effect with a `(sessionId, turnSeq)` token: the intent is "already pressed Stop for this exact session+turn", so a session switch (or a new turn) is automatically a mismatch and the next first Stop is graceful again. No cross-session leak, and no prop-driven effect.
